src/daemon/daemon.cpp

```diff
@@ -678,26 +763,48 @@ mp::Daemon::Daemon(std::unique_ptr<const DaemonConfig> the_config)
 {
     connect_rpc(daemon_rpc, *this);
     std::vector<std::string> invalid_specs;
-    bool mac_addr_missing{false};
+
     for (auto& entry : vm_instance_specs)
     {
         const auto& name = entry.first;
         auto& spec = entry.second;
 
+        // Store here the MAC addresses we need add to the daemon's set. But we'll add them only if this instance
+        // is not invalid.
+        std::unordered_set<std::string> instance_mac_addresses;
+
         if (!config->vault->has_record_for(name))
         {
             invalid_specs.push_back(name);
             continue;
         }
 
-        auto mac_addr = spec.mac_addr;
-        if (mac_addr.empty())
+        // Check MAC addresses of the default and extra interfaces. For each MAC, repetition on the instance MAC
+        // addresses and in the daemon's MAC addresses is checked. String validity was already checked in
+        // load_db().
+        if (!mpu::valid_mac_address(spec.default_interface.mac_address) ||
+            !instance_mac_addresses.insert(spec.default_interface.mac_address).second ||
+            allocated_mac_addrs.find(spec.default_interface.mac_address) != allocated_mac_addrs.end())
```

Comment:
A few points here:

- IIUC, your comment implies that there's no need to recheck string validity here, which I agree with. But then you forgot to remove the check from the code, right?
- No point checking for insertion success here, because the only way that bool could be false is if the mac was already there, which couldn't happen since the set was just created. Note: unfortunately, cppreference is not very clear on the meaning of the bool that insert returns, but the standard says: 

    > "Inserts t if and only if there is no element in the container with key equivalent to the key of t. The bool component of the returned pair is true if and only if the insertion takes place [...]". 

    Which implies "false *iff* key there already". 

- finally, no point inserting into `instance_mac_addresses` if it is already in `allocated_mac_addrs` (although this is more relevant in the analogous diff below).

Combining all that, I think this could be written as (didn't test or compile):
```suggestion
        // Check MAC addresses of the default and extra interfaces. For each MAC, repetition on the instance MAC
        // addresses and in the daemon's MAC addresses is checked. String validity was already checked in
        // load_db().
        if (allocated_mac_addrs.find(spec.default_interface.mac_address) == allocated_mac_addrs.end())
            instance_mac_addresses.insert(spec.default_interface.mac_address)
        else
```

Reply:
1. Yes, the comment is right but the code checks the string, the check needs to be removed.
2. There must be two checks for repetition: it should be checked (a) if the MAC is on another instance and (b) if the MAC was already specified on this instance. That's why I put insert, we need the value to be inserted and we need to know if it was already there: we can do both thing with one call to `insert`.
So I removed the validity check but I left the rest.

Reply:
1. Ack
2. `instance_mac_addresses` was empty at this point, so no "intra-instance" repetition would be detected here. Anyway, with other changes this is not relevant anymore, so resolving.